triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -614,14 +627,43 @@ private[lf] class Runner private (
     var numberOfExerciseByKey = 0
 
     val startTime = System.nanoTime()
+    val ruleEvaluationTimer = materializer.scheduleOnce(
+      triggerConfig.hardLimit.ruleEvaluationTimeout,
+      () => {
+        triggerContext.logError(
+          "Stopping trigger as the rule evaluator as exceeded its allotted running time",
+          "rule-evaluation-timeout" -> triggerConfig.hardLimit.ruleEvaluationTimeout,
+        )
+
+        throw TriggerRuleEvaluationTimeout(triggerConfig.hardLimit.ruleEvaluationTimeout)
+      },
+    )
 
     triggerContext.childSpan("step") { implicit triggerContext: TriggerLogContext =>
       def evaluate(se: SExpr): SValue = {
+        val stepIteratorTimer = materializer.scheduleOnce(
+          triggerConfig.hardLimit.stepInterpreterTimeout,
+          () => {
+            triggerContext.logError(
+              "Stopping trigger as the rule step interpreter as exceeded its allotted running time",
+              "step-interpreter-timeout" -> triggerConfig.hardLimit.stepInterpreterTimeout,
+            )
+
+            throw TriggerRuleStepInterpretationTimeout(
+              triggerConfig.hardLimit.stepInterpreterTimeout
+            )
+          },
+        )
         val machine: Speedy.PureMachine =
           Speedy.Machine.fromPureSExpr(compiledPackages, se)
+
         // Evaluate it.
-        machine.setExpressionToEvaluate(se)
-        Machine.stepToValue(machine)
+        try {
+          machine.setExpressionToEvaluate(se)
+          Machine.stepToValue(machine)
+        } finally {
+          discard(stepIteratorTimer.cancel())
```

Comment:
Using `finally` as we need to ensure the timer is stopped (regardless of whether the try block succeeds or throws). If the timer can not be cancelled, then something _bad_ is happening(?) and the trigger will be stopped (because the exception will be thrown).